Reduced compilation time

fbridault opened the pull request:
At last, reduced drastically (from 40% to 55%) the compilation time.
- the big gain comes by precompiled headers usage. Most of them have been customized but we can probably do even better for some of them.
- a second gain comes from a trick to compile camp macros only when needed.
- a minor gain comes from some cleaning in headers includes.

There is a CMake option to switch off the pch support, which is on by default. Everything is reported to work well on Linux (gcc/clang) and Windows. MacOsX support might work since clang is handled. If necessary I can disable it by default on MacOSX if it does not work. It would be sad to not benefit from this huge gain on the other platforms.

fbridault committed:
Removed camp headers and macros for most compile units.

I kept them only when compiling fwDataCamp, fwMedDataCamp and fwAtomsConversion.

fbridault committed:
Removed unecessary inclusions of fwData/Object.hpp

No performance gain ofc, this is just for clarity.

fbridault committed:
Added draft of pch support, currently only for clang.

Added a pch for fwData : 75% faster.

fbridault committed:
Added pchs for SrcLib and Bundles.

fbridault committed:
Customized some pchs and removed some others that decrease performance.

For instance when there is only one source file in the library, the use of a
pch causes a slower build.

fbridault committed:
Added basic MSVC support for pch.

fbridault committed:
Fixed order of includes in fwGdcmIO and ioGdcm pch for windows.

fbridault committed:
Changed cmake pch message mode to status.

fbridault committed:
Removed the necessity to use add_definitions before fwLoadProperties

I also tried to use target_compile_definitions instead of add_definitions whenever possible. This allows us to benefit from custom interfaces, i.e. to define public defines.

fbridault committed:
Added an option to trigger pch use.

fbridault committed:
Added GCC support for pch

fbridault committed:
Fixed missing newSlot symbol.

fbridault commented:
Commits have been rebased and reordered/mergered. Please discard your local checkout if you tried it before.

emiliehar commented on CMake/modules/PrecompiledHeader.cmake:
```diff
@@ -0,0 +1,249 @@
+    endif()
+  endif(MSVC)
+
+  if(CMAKE_COMPILER_IS_GNUCXX OR "${CMAKE_CXX_COMPILER_ID}" STREQUAL "Clang")
```
On MacOS, the CMAKE_CXX_COMPILER_ID is "AppleClang". 
We should use MATCHES instead of STREQUAL

It works using:
   `if(CMAKE_COMPILER_IS_GNUCXX OR "${CMAKE_CXX_COMPILER_ID}" MATCHES "Clang")`

emiliehar commented on CMake/modules/PrecompiledHeader.cmake:
```diff
@@ -0,0 +1,249 @@
+    set(_pch_binary_dir "${CMAKE_CURRENT_BINARY_DIR}")
+    set(_pchfile "${_pch_binary_dir}/${_input}")
+
+    if("${CMAKE_CXX_COMPILER_ID}" STREQUAL "Clang")
```
Use MATCHES instead of STREQUAL

emiliehar commented on CMake/modules/PrecompiledHeader.cmake:
```diff
@@ -0,0 +1,249 @@
+        endif()
+        separate_arguments(_pch_compile_flags)
+        list(APPEND _pch_compile_flags -Winvalid-pch)
+        if("${CMAKE_CXX_COMPILER_ID}" STREQUAL "Clang")
```
Same, use MATCHES instead of STREQUAL

emiliehar reviewed (changes requested):
With the little modifications in PrecompiledHeader.cmake, my compilation time of VRRender decrease from 12 minutes to less than 8 minutes.
Test on MacOS El Capitan with Xcode 8.2

fbridault committed:
Fixed Clang detection in PrecompiledHeader.cmake on MacOsX.

fbridault commented:
Thanks for your tests and remarks @emiliehar I integrated your changes. :+1:

fbridault committed:
Fixed pch dependencies with gcc and clang.

fbridault committed:
Simplified dependencies scanning with gcc and clang.

fbridault committed:
Added two default pchs: one for libraries and one for bundles.

fbridault committed:
Added pch sharing between targets.

fbridault committed:
Fixed dependency trouble with moc and pch both accessing vc120.pdb on Windows.

fbridault committed:
Removed obsolete ::fwTools::DynamicAttributes class and files.

fbridault committed:
Fixed dependencies issues with pch sharing.

 Also, added two new bundles to contain the default pch for libs and bundles.

fbridault committed:
Customized pch a bit more.

fbridault committed:
Fixed gcc pch build.

fbridault commented:
I made several improvements in the last commits.

I fixed a bug with pch dependencies with Clang and Gcc. If a file included in a pch was modified, then the pch was not automatically recompiled by Ninja or Makefile. It could have been fixed easily with the `IMPLICIT_DEPENDS` option of add_custom_command but this is not implemented for the Ninja generator. Thus I chose a different approach, using the `DEPFILE` option of `add_custom_command`. The dependency file is generated in a custom command before compiling the pch.

In the last commits, I improved the overall pch usage. I implemented pch sharing, so that it is now possible to use a pch in a bundle or a library from an another target by setting `USE_PCH_FROM_TARGET` in the `Properties.cmake`. There are also two new dummy targets, `pchCore` and `pchServices`, that allow to compile and use a pch with a good base for core libraries or a component using services. `pchCore` is used by default for libraries and `pchServices` for bundles, if no custom `pch.hpp` is found and `USE_PCH_FROM_TARGET` is not set. This sharing strategy allowed to gain around 15% more. :) And the nice thing is that even if you don't customize your pch, you already get a good boost with the default pch.
I also took care to manage dependencies correctly, so if you specify a pch target that was not already a dependency or a requirement, then the build script automatically adds this target as a requirement.

Gcc support forced me to add a constraint that was not present before. It is not allowed now to include files from the current target in a pch. This is to simplify pch sharing: when using a pch from an another target, the cpp file must include the same definitions than the pch. So I removed `target_EXPORTS` (and `target_VER` but this is less important) definitions from the pch build. This prevents to export symbols from the pch included file. Practically this is not such a big deal, if you would like to include a header from the current target, just try to copy the bigger included files outside of the target from this header. I did this for the pch where it occurred and I noticed no performance penalty.

Nevertheless I have to test all these improvements on Windows and then, I think we are good for the merge.

fbridault committed:
Implemented Windows pch share.

fbridault committed:
Added pchData for libraries or bundles depending mainly on fwData.

fbridault committed:
Cleaned some headers to reduce compile time.

fbridault committed:
Added some comments in PrecompiledHeader.cmake

fbridault committed:
Removed most of pch.

After lots of benchmarks, results are not better than using default pch.

fbridault committed:
Changed pch bundles into libraries.

fbridault commented:
Okay I think I'm ready for the last review round. I fixed the windows build, it was harder than expected, especially because of the need to link with the original pch.obj when sharing pch. I solved that issue thanks to CMake "Object" library. 
I removed most custom pch because the benchmarks prove they don't improve performance and depending on the platform, on windows for instance, building a pch is a bottleneck since Ninja breaks the parallel build at this stage. It so better to keep the number of pch quite low. 
This sharing strategy is also very efficient on Windows. On my windows laptop, I started with 23 minutes to build VRRender. With the first pch implementation without sharing, I managed to build in 11 minutes. With pch sharing, I managed to build even faster in 6 minutes.

Please test and review the code, I think this would be a very nice addition for 11.0.4.

fbridault committed:
Fixed issue on macOS with parsing sed arguments.

fbridault committed:
Added CMake minimum version requirement.

schweitzer commented:
Ok it works well for me!

Here's the details of my tests:

# Linux
* **OS**: Linux Mint 18.1
* **Processor**: Intel(R) Core(TM) i7-5930K CPU @ 3.50GHz (12 cores)
* **RAM**: 16go
* **Compiler** : GCC 5.4
* **CMake**: 3.7.2
* **Build Tool**: ninja (with `-j10` flag)
* **Build Type** : Debug (with Tests)

## Time to build VRRender from scratch:
**Before PCH**:
`real: 11m16.126s`
`user: 104m8.964s`
`sys: 6m16.620s`
**After PCH**:
`real: 5m55.440s`
`user: 51m20.940s`
`sys: 3m46.300s`

It is **2x** faster using PCH!

# Windows
* **OS**: Windows 8.1
* **Processor**: Intel(R) Core(TM) i7 CPU M620 @2.67GHZ (4 cores)
* **RAM**: 6go
* **Compiler** : msvs 2013
* **CMake**: 3.7.0
* **Build Tool**: ninja (with `-j4` flag)
* **Build Type** : Debug (with Tests)

## Time to build VRRender from scratch:
**Before PCH**:
*68min*
**After PCH**:
*27min*

It is at least **2x** faster with PCH!

Great work !

dweckmann commented:
On my machine:

macOS SIERRA 10.12.3 (16D32)
Mac Pro (mi-2010)
2 x 2,4 GHz Quad-Core Intel Xeon
24 Go 1066 MHz DDR3

The result are also impressive:

Without PCH:
real    15m29.658s
user    221m4.906s
sys      11m24.929s

Witch PCH:
real    8m27.573s
user   114m31.416s
sys     7m25.137s

fbridault committed:
Updated .travis.yml to avoid the cmake 3.7 requirement of the PCH.

greenjava commented:
### Configuration
- **OS:** Windows 10 Version 1607
- **Processor:** Intel(R) Core(TM) i7 CPU 5930K @3.50GHZ (12 cores)
- **RAM:** 16go
- **Compiler:** MSVC 2015 Update 3
- **CMake:** 3.7.0
- **Build Tool:** ninja
- **Build type:** Debug (without test)

### Time to build VRRender from scratch:
**Before PCH:**
12min14.751s
**After PCH:**
6min5.621s

👍

fbridault committed:
Fixed cmake configure when a component is skipped in the build.

fbridault committed:
Fixed spylog level macros override.

It was quite hard to fix, but I managed to keep usage identical.

fbridault committed:
Merge branch 'dev'

Conflicts:
	Bundles/LeafCtrl/ctrlSelection/include/ctrlSelection/updater/SObjFromUid.hpp

fbridault commented:
Okay, I truly hope this was the very last commit for this pull request. 

I realized that the `SPYLOG_LEVEL_<target>` override was broken. It was not possible to define a different level than the one defined in the pch target.

I fixed this on the three different platforms, it was really a pain in the ass... There is a little difference than before, log macros defined in hpp will use the log level from the pch and not the target. Actually I think it was a bug before. When we defined the log level `TRACE` in `ioVTK` for instance, the traces from log macros of `fwCom` headers were displayed. So now you will only get the traces from `ioVTK`, when using pch. Without the pch the behaviour remains the same.
